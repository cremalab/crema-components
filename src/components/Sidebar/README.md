# `<Sidebar />`

DESCRIPTION_HERE

## Directory Structure

- `Sidebar.stories.tsx`: Component stories (`npm run test:playground`)
- `Sidebar.test.tsx`: Component tests (`npm run test:unit`)
- `Sidebar.tsx`: Component code
- `index.ts`: Component export
- `README.md`: Component documentation (hey, that's me!)
