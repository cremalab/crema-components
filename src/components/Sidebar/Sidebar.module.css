.sidebar {
  display: flex;
  flex-direction: column;
  background: #f2f3f5;
  width: 300px;
  height: 100%;
  position: fixed;
  top: 0;
  padding: 1rem 1.5rem;
  z-index: 10;
  box-shadow: 0 0 5px #f2f3f5;
}

.sidebarOverlay {
  background-color: hsla(170, 5%, 15%, 0.8);
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.sidebarRight {
  right: 0;
}

.sidebarTitle {
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
}

.sidebarTitleClose {
  color: #808080;
  cursor: pointer;
}