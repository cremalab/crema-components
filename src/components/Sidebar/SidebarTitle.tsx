import { ReactNode } from "react"
import { Button } from "../Button"
import styles from "./Sidebar.module.css"

interface Props {
  children: ReactNode
  onClose?: () => unknown
}

export function SidebarTitle({ children, onClose }: Props) {
  return (
    <div className={styles.sidebarTitle}>
      <div>{typeof children === "string" ? <p>{children}</p> : children}</div>
      {onClose ? (
        <div>
          {
            <Button
              className={styles.sidebarTitleClose}
              ariaLabel="close the sidebar"
              onClick={onClose}
            >
              x
            </Button>
          }
        </div>
      ) : null}
    </div>
  )
}
