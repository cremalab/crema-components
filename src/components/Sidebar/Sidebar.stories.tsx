import { ComponentMeta, Story } from "@storybook/react"
import { ComponentProps, useState } from "react"
import { Button } from "../Button"
import { TextInput } from "../TextInput"
import { Sidebar } from "./Sidebar"
import styles from "./Sidebar.module.css"
/**
 * See Storybook Docs: Writing Stories
 * https://storybook.js.org/docs/basics/writing-stories/
 */

export default {
  title: "Components/Sidebar",
  component: Sidebar,
} as ComponentMeta<typeof Sidebar>

export const Basic: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open basic sidebar
      </Button>
      <Sidebar
        open={sidebarExpanded}
        hideOverlay={false}
        position="left"
        onClose={() => setSidebarExpanded(false)}
      >
        <div className={styles.sidebarChildren}>
          <p>I am a basic sidebar ðŸ˜ƒ</p>
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}

export const WithTitle: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open sidebar with title
      </Button>
      <Sidebar
        title="Hello from the sidebar!"
        hideOverlay={false}
        open={sidebarExpanded}
        onClose={() => setSidebarExpanded(false)}
      >
        <div className={styles.sidebarChildren}>
          <p>I am a sidebar. I have a title! ðŸ˜ƒ</p>
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}

export const WithOverlay: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open sidebar with overlay
      </Button>
      <Sidebar
        title="Sidebar With Title"
        open={sidebarExpanded}
        onClose={() => setSidebarExpanded(false)}
        hideOverlay={false}
      >
        <div className={styles.sidebarChildren}>
          <p>Sidebars can be cool</p>
          <p>I am a sidebar. I have a title! ðŸ˜ƒ</p>
          <TextInput name="test-input" label="Test Input" hideLabel={false} />
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}

export const NoOverlay: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open sidebar without overlay
      </Button>
      <Sidebar
        title="Sidebar With Title"
        open={sidebarExpanded}
        onClose={() => setSidebarExpanded(false)}
        hideOverlay={true}
      >
        <div className={styles.sidebarChildren}>
          <p>Sidebars can be cool</p>
          <p>I am a sidebar. I have a title! ðŸ˜ƒ</p>
          <TextInput name="test-input" label="Test Input" hideLabel={false} />
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}

export const LeftPosition: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open sidebar on the left
      </Button>
      <Sidebar
        title="Sidebar With Title"
        open={sidebarExpanded}
        onClose={() => setSidebarExpanded(false)}
        position="left"
      >
        <div className={styles.sidebarChildren}>
          <p>Sidebars can be cool</p>
          <p>I am a sidebar. I have a title! ðŸ˜ƒ</p>
          <TextInput name="test-input" label="Test Input" hideLabel={false} />
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}

export const RightPosition: Story<ComponentProps<typeof Sidebar>> = () => {
  const [sidebarExpanded, setSidebarExpanded] = useState(false)

  return (
    <div>
      <Button ariaLabel="Open" onClick={() => setSidebarExpanded(true)}>
        Open sidebar on the right
      </Button>
      <Sidebar
        title="Sidebar With Title"
        open={sidebarExpanded}
        onClose={() => setSidebarExpanded(false)}
        position="right"
      >
        <div className={styles.sidebarChildren}>
          <p>Sidebars can be cool</p>
          <p>I am a sidebar. I have a title! ðŸ˜ƒ</p>
          <TextInput name="test-input" label="Test Input" hideLabel={false} />
        </div>
        <Button
          ariaLabel="Close Sidebar"
          aria-labelledby="close-sidebar"
          className="close-sidebar"
          onClick={() => setSidebarExpanded(false)}
        >
          Close
        </Button>
      </Sidebar>
    </div>
  )
}
