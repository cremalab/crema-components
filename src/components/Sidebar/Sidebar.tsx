import classNames from "classnames"
import { HTMLProps, ReactNode, useRef } from "react"
import { createPortal } from "react-dom"
import { useKeyPress } from "../../hooks/useKeyPress"
import { useOutsideClick } from "../../hooks/useOutsideClick"
import styles from "./Sidebar.module.css"
import { SidebarTitle } from "./SidebarTitle"

export interface Props extends HTMLProps<HTMLDivElement> {
  children: ReactNode
  hideOverlay?: boolean
  onClose?: () => void
  position?: "left" | "right"
}

export const Sidebar = ({
  children,
  hideOverlay,
  open,
  onClose,
  position = "left",
  title,
  ...props
}: Props) => {
  const sidebarRef = useRef<HTMLDivElement>(null)

  useOutsideClick(sidebarRef, () => {
    onClose?.()
  })

  useKeyPress(["Escape"], () => {
    onClose?.()
  })

  if (!open) return null

  return createPortal(
    <>
      {!hideOverlay && (
        <div
          className={styles.sidebarOverlay}
          onClick={onClose}
          data-testid="sidebar-backdrop"
        ></div>
      )}
      <div
        className={classNames(styles.sidebar, {
          [styles.sidebarRight]: position === "right",
          [styles.sidebarLeft]: position === "left",
        })}
        ref={sidebarRef}
        {...props}
      >
        {title && <SidebarTitle onClose={onClose}>{title}</SidebarTitle>}
        {children}
      </div>
    </>,
    document.body,
  )
}
