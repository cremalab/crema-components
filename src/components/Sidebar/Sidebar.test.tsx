import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Sidebar } from "./Sidebar"
import { SidebarTitle } from "./SidebarTitle"

describe("Sidebar", () => {
  it("renders nothing when open = false", async () => {
    // Arrange
    const children = "This should not render"

    // Act
    render(<Sidebar open={false}>{children}</Sidebar>)

    // Assert
    expect(screen.queryByText(children)).toBeNull()
  })

  it("renders content when open = true", async () => {
    // Arrange
    const title = "title"
    const children = "children"

    // Act
    render(
      <Sidebar open={true} title={title}>
        {children}
      </Sidebar>,
    )

    const titleNode = screen.queryByText(title)
    const childrenNode = screen.queryByText(children)

    // Assert
    expect(titleNode).toBeInTheDocument()
    expect(childrenNode).toBeInTheDocument()
  })

  it("calls onClose when the sidebar overlay is clicked", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Sidebar open={true} title={title} onClose={onClose} hideOverlay={false}>
        {children}
      </Sidebar>,
    )

    const overlay = await screen.findByTestId("sidebar-backdrop")
    await userEvent.click(overlay)

    expect(onClose).toHaveBeenCalled()
  })

  it("calls onClose when the escape key is pressed", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Sidebar open={true} title={title} onClose={onClose}>
        {children}
      </Sidebar>,
    )

    await userEvent.keyboard("[Escape]")

    expect(onClose).toHaveBeenCalled()
  })

  it("does not call onClose when the sidebar is not open and the escape key is pressed", async () => {
    // Arrange
    const onClose = jest.fn()
    const title = "title"
    const children = "children"

    // Act
    render(
      <Sidebar open={false} title={title} onClose={onClose}>
        {children}
      </Sidebar>,
    )

    await userEvent.keyboard("[Escape]")

    expect(onClose).not.toHaveBeenCalled()
  })

  it("does not render a title when title is not provided", async () => {
    // Arrange
    const children = "children"

    // Act
    render(<Sidebar open={true}>{children}</Sidebar>)

    // Assert
    expect(screen.getByText(children)).toBeInTheDocument()
  })
})

describe("SidebarTitle", () => {
  it("renders the children", async () => {
    // Arrange
    // Act
    render(<SidebarTitle>Test Title</SidebarTitle>)
    const received = screen.queryByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
  })

  it("does not render the close button if no onClose is given", async () => {
    // Arrange
    // Act
    render(<SidebarTitle>Test Title</SidebarTitle>)

    const closeButton = screen.queryByText("close the sidebar")

    // Assert
    expect(closeButton).toBeNull()
  })

  it("calls onClose when the close button is clicked", async () => {
    // Arrange
    const onClose = jest.fn()

    // Act
    render(<SidebarTitle onClose={onClose}>Test Title</SidebarTitle>)
    const closeButton = screen.getByRole("button")
    await userEvent.click(closeButton)

    // Assert
    expect(onClose).toHaveBeenCalledTimes(1)
  })

  it("renders children in a paragraph tag if children is a string", async () => {
    // Arrange
    const children = "Test Title"
    // Act
    render(<SidebarTitle>{children}</SidebarTitle>)
    const received = screen.queryByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
    expect(received?.tagName).toBe("P")
  })

  it("renders children in a div tag if children is not a string", async () => {
    // Arrange
    const children = <div>Test Title</div>
    // Act
    render(<SidebarTitle>{children}</SidebarTitle>)
    const received = screen.queryByText("Test Title")

    // Assert
    expect(received).toBeInTheDocument()
    expect(received?.tagName).toBe("DIV")
  })
})
