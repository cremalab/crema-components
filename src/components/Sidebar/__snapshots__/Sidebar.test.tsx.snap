// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders nothing when open = false 1`] = `<DocumentFragment />`;
